### PR TITLE
Feature/release 2.5

### PR DESCRIPTION
**Pull request by hardik-desai-kcl**

### CODE REVIEW

File: src/Sfa.Tl.Matching.Application.IntegrationTests/MatchingServiceReport/Builders/OpportunityBuilder.cs [hunk 1]
```diff
@@ -0,0 +1,252 @@
...
+            };
+
+            _context.Add(opportunity);
+
+            _context.SaveChanges();
+
+            return opportunity;
+        }
+
+        public OpportunityItem CreaeReferralOpportunityItem(bool isSaved, bool isCompleted, params int[] providerVenueId)
```
Comment (adamesfa) [inline src/Sfa.Tl.Matching.Application.IntegrationTests/MatchingServiceReport/Builders/OpportunityBuilder.cs]: Should be CreateReferral...

File: src/Sfa.Tl.Matching.Application.IntegrationTests/MatchingServiceReport/Builders/OpportunityBuilder.cs [hunk 2]
```diff
@@ -0,0 +1,252 @@
...
+    public class ProviderBuilder
+    {
+        private readonly MatchingDbContext _context;
+
+        public ProviderBuilder(MatchingDbContext context)
+        {
+            _context = context;
+        }
+
+        public Provider CreaeProvider(int venueCount = 1)
```
Comment (adamesfa) [inline src/Sfa.Tl.Matching.Application.IntegrationTests/MatchingServiceReport/Builders/OpportunityBuilder.cs]: Should be CreateProvider

File: src/Sfa.Tl.Matching.Application.IntegrationTests/MatchingServiceReport/Builders/OpportunityBuilder.cs [hunk 3]
```diff
@@ -0,0 +1,252 @@
...
+    public class EmployerBuilder
+    {
+        private readonly MatchingDbContext _context;
+
+        public EmployerBuilder(MatchingDbContext context)
+        {
+            _context = context;
+        }
+
+        public Domain.Models.Employer CreaeEmployer(Guid crmId)
```
Comment (adamesfa) [inline src/Sfa.Tl.Matching.Application.IntegrationTests/MatchingServiceReport/Builders/OpportunityBuilder.cs]: Should be CreateEmployer

File: src/Sfa.Tl.Matching.Application.IntegrationTests/MatchingServiceReport/MatchingServiceProviderEmployerReport/When_MatchingServiceProviderEmployerReport_Is_Called_For_UnCompleted_Referral.cs [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+        {
+            _dbContext = new TestConfiguration().GetDbContext();
+
+            _opportunityBuilder = new OpportunityBuilder(_dbContext);
+            _providerBuilder = new ProviderBuilder(_dbContext);
+            _employerBuilder = new EmployerBuilder(_dbContext);
+
+            ClearData();
+
+            var employer = _employerBuilder.CreaeEmployer(Guid.NewGuid());
```
Comment (adamesfa) [inline src/Sfa.Tl.Matching.Application.IntegrationTests/MatchingServiceReport/MatchingServiceProviderEmployerReport/When_MatchingServiceProviderEmployerReport_Is_Called_For_UnCompleted_Referral.cs]: Same here

File: src/Sfa.Tl.Matching.Application.IntegrationTests/MatchingServiceReport/MatchingServiceProviderEmployerReport/When_MatchingServiceProviderEmployerReport_Is_Called_For_UnCompleted_Referral.cs [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+            var provider = _providerBuilder.CreaeProvider();
```
Comment (adamesfa) [inline src/Sfa.Tl.Matching.Application.IntegrationTests/MatchingServiceReport/MatchingServiceProviderEmployerReport/When_MatchingServiceProviderEmployerReport_Is_Called_For_UnCompleted_Referral.cs]: Same here

File: src/Sfa.Tl.Matching.Application.IntegrationTests/MatchingServiceReport/MatchingServiceProviderEmployerReport/When_MatchingServiceProviderEmployerReport_Is_Called_For_UnCompleted_Referral.cs [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+            
+            _savedReferralOpportunityItem = _opportunityBuilder.CreaeReferralOpportunityItem(true, false, provider.ProviderVenue.First().Id);
```
Comment (adamesfa) [inline src/Sfa.Tl.Matching.Application.IntegrationTests/MatchingServiceReport/MatchingServiceProviderEmployerReport/When_MatchingServiceProviderEmployerReport_Is_Called_For_UnCompleted_Referral.cs]: Same here

File: src/Sfa.Tl.Matching.Application.IntegrationTests/MatchingServiceReport/Builders/OpportunityBuilder.cs (hunk 2 above)
Comment (nandakumar0311) [inline src/Sfa.Tl.Matching.Application.IntegrationTests/MatchingServiceReport/Builders/OpportunityBuilder.cs]: CreateProvider, please?

File: src/Sfa.Tl.Matching.Application.IntegrationTests/MatchingServiceReport/MatchingServiceOpportunityReport/When_MatchingServiceOpportunityReport_Is_Called_For_Pipeline_Opportunity.cs [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+            _savedReferralOpportunityItem = _opportunityBuilder.CreaeReferralOpportunityItem(true, false, provider.ProviderVenue.First().Id);
+            _savedProvisionGapOpportunityItem = _opportunityBuilder.CreateProvisionGapOpportunityItem(true, false);
+
+            _opportunityBuilder.CreateOpportunity(employer.CrmId, new List<OpportunityItem> { _savedReferralOpportunityItem, _savedProvisionGapOpportunityItem });
+
+            _result = _dbContext.MatchingServiceOpportunityReport.ToList();
+        }
+
+        [Fact]
+        public void Then_Pipeline_referal_Is_included()
```
Comment (adamesfa) [inline src/Sfa.Tl.Matching.Application.IntegrationTests/MatchingServiceReport/MatchingServiceOpportunityReport/When_MatchingServiceOpportunityReport_Is_Called_For_Pipeline_Opportunity.cs]: Should be Then_Pipeline_Referral_Is_Included

File: src/Sfa.Tl.Matching.Application.IntegrationTests/MatchingServiceReport/MatchingServiceOpportunityReport/When_MatchingServiceOpportunityReport_Is_Called_For_Saved_OpportunityOpportunities.cs [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+            _unsavedReferralOpportunityItem = _opportunityBuilder.CreaeReferralOpportunityItem(false, false, provider2.ProviderVenue.First().Id);
+            _unsavedProvisionGapOpportunityItem = _opportunityBuilder.CreateProvisionGapOpportunityItem(false, false);
+
+            _opportunityBuilder.CreateOpportunity(employer.CrmId, new List<OpportunityItem> { _savedReferralOpportunityItem, _savedProvisionGapOpportunityItem, _unsavedReferralOpportunityItem, _unsavedProvisionGapOpportunityItem });
+
+            _result = _dbContext.MatchingServiceOpportunityReport.ToList();
+        }
+
+        [Fact]
+        public void Then_Save_referal_Is_included()
```
Comment (adamesfa) [inline src/Sfa.Tl.Matching.Application.IntegrationTests/MatchingServiceReport/MatchingServiceOpportunityReport/When_MatchingServiceOpportunityReport_Is_Called_For_Saved_OpportunityOpportunities.cs]: Should be Then_Save_Referral_Is_Included

File: src/Sfa.Tl.Matching.Application.IntegrationTests/MatchingServiceReport/MatchingServiceOpportunityReport/When_MatchingServiceOpportunityReport_Is_Called_For_Saved_OpportunityOpportunities.cs [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+
+        [Fact]
+        public void Then_Saved_ProvisionGap_Is_Included()
+        {
+            _result.Should().NotBeNull();
+            _result.SingleOrDefault(o => o.OpportunityItemId == _savedProvisionGapOpportunityItem.Id).Should().NotBeNull();
+        }
+
+        [Fact]
+        public void Then_Unsave_referal_Is_NOT_included()
```
Comment (adamesfa) [inline src/Sfa.Tl.Matching.Application.IntegrationTests/MatchingServiceReport/MatchingServiceOpportunityReport/When_MatchingServiceOpportunityReport_Is_Called_For_Saved_OpportunityOpportunities.cs]: Should be Then_Unsaved_Referral_Is_NOT_Included

File: src/Sfa.Tl.Matching.Application.IntegrationTests/MatchingServiceReport/Builders/OpportunityBuilder.cs (hunk 2 above)
Comment (mikeymonster) [reply src/Sfa.Tl.Matching.Application.IntegrationTests/MatchingServiceReport/Builders/OpportunityBuilder.cs]: Should there be 3 files for the 3 classes in this file?

File: src/Sfa.Tl.Matching.Application.IntegrationTests/MatchingServiceReport/MatchingServiceProviderEmployerReport/When_MatchingServiceProviderEmployerReport_Is_Called_For_Saved_And_Completed_Referral_And_ProvisionGap.cs [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+
+        private void ClearData()
+        {
+            _opportunityBuilder.ClearData();
+            _providerBuilder.ClearData();
+            _employerBuilder.ClearData();
+        }
+
+        [Fact]
+        public void Then_provisionGap_Is_Not_Selected()
```
Comment (adamesfa) [inline src/Sfa.Tl.Matching.Application.IntegrationTests/MatchingServiceReport/MatchingServiceProviderEmployerReport/When_MatchingServiceProviderEmployerReport_Is_Called_For_Saved_And_Completed_Referral_And_ProvisionGap.cs]: Should be Then_ProvisionGap_Is_Not_Selected

File: src/Sfa.Tl.Matching.Application.IntegrationTests/MatchingServiceReport/MatchingServiceProviderEmployerReport/When_MatchingServiceProviderEmployerReport_Is_Called_For_UnCompleted_Referral.cs (hunk 4 above)
Comment (adamesfa) [inline src/Sfa.Tl.Matching.Application.IntegrationTests/MatchingServiceReport/MatchingServiceProviderEmployerReport/When_MatchingServiceProviderEmployerReport_Is_Called_For_UnCompleted_Referral.cs]: Same here

File: src/Sfa.Tl.Matching.Application.IntegrationTests/MatchingServiceReport/MatchingServiceOpportunityReport/When_MatchingServiceOpportunityReport_Is_Called_For_Pipeline_Opportunity.cs (hunk 7 above)
Comment (mikeymonster) [inline src/Sfa.Tl.Matching.Application.IntegrationTests/MatchingServiceReport/MatchingServiceOpportunityReport/When_MatchingServiceOpportunityReport_Is_Called_For_Pipeline_Opportunity.cs]: spelling - `_Referral_` and capital I on Included

File: src/Sfa.Tl.Matching.Application.IntegrationTests/MatchingServiceReport/MatchingServiceOpportunityReport/When_MatchingServiceOpportunityReport_Is_Called_For_Saved_OpportunityOpportunities.cs (hunk 8 above)
Comment (mikeymonster) [inline src/Sfa.Tl.Matching.Application.IntegrationTests/MatchingServiceReport/MatchingServiceOpportunityReport/When_MatchingServiceOpportunityReport_Is_Called_For_Saved_OpportunityOpportunities.cs]: spelling - `_Referral_` and capital I on Included

File: src/Sfa.Tl.Matching.Application.IntegrationTests/MatchingServiceReport/MatchingServiceOpportunityReport/When_MatchingServiceOpportunityReport_Is_Called_For_Saved_OpportunityOpportunities.cs [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+        private readonly OpportunityItem _savedProvisionGapOpportunityItem;
+        private readonly OpportunityItem _savedReferralOpportunityItem;
+        private readonly OpportunityItem _unsavedProvisionGapOpportunityItem;
+        private readonly OpportunityItem _unsavedReferralOpportunityItem;
+
+        private readonly OpportunityBuilder _opportunityBuilder;
+        private readonly ProviderBuilder _providerBuilder;
+        private readonly EmployerBuilder _employerBuilder;
+
+        public When_MatchingServiceOpportunityReport_Is_Called_For_Saved_OpportunityOpportunities()
```
Comment (mikeymonster) [inline src/Sfa.Tl.Matching.Application.IntegrationTests/MatchingServiceReport/MatchingServiceOpportunityReport/When_MatchingServiceOpportunityReport_Is_Called_For_Saved_OpportunityOpportunities.cs]: Not sure about this name - what's are "Opportunity Opportunities"?

File: src/Sfa.Tl.Matching.Application.IntegrationTests/MatchingServiceReport/MatchingServiceOpportunityReport/When_MatchingServiceOpportunityReport_Is_Called_For_Saved_OpportunityOpportunities.cs (hunk 9 above)
Comment (mikeymonster) [inline src/Sfa.Tl.Matching.Application.IntegrationTests/MatchingServiceReport/MatchingServiceOpportunityReport/When_MatchingServiceOpportunityReport_Is_Called_For_Saved_OpportunityOpportunities.cs]: spelling - `_Referral_` and capital I on Included

File: src/Sfa.Tl.Matching.Application.IntegrationTests/MatchingServiceReport/MatchingServiceProviderEmployerReport/When_MatchingServiceProviderEmployerReport_Is_Called_For_Saved_And_Completed_Referral_And_ProvisionGap.cs (hunk 10 above)
Comment (mikeymonster) [inline src/Sfa.Tl.Matching.Application.IntegrationTests/MatchingServiceReport/MatchingServiceProviderEmployerReport/When_MatchingServiceProviderEmployerReport_Is_Called_For_Saved_And_Completed_Referral_And_ProvisionGap.cs]: Capital P on ProvisionGap

File: src/Sfa.Tl.Matching.Application.IntegrationTests/MatchingServiceReport/MatchingServiceProviderEmployerReport/When_MatchingServiceProviderEmployerReport_Is_Called_For_UnCompleted_Referral.cs (hunk 4 above)
Comment (hardik-desai-kcl) [reply src/Sfa.Tl.Matching.Application.IntegrationTests/MatchingServiceReport/MatchingServiceProviderEmployerReport/When_MatchingServiceProviderEmployerReport_Is_Called_For_UnCompleted_Referral.cs]: Done

**Pull request merged**